### PR TITLE
Add support for Quantization-aware search on OFA ResNet50 supernet

### PR DESCRIPTION
**Pull request by macsz**
This PR enables you to run search on your FP32 super-network and find optimal model configurations w.r.t. both architecture and Post-Training Quantization policy. DyNAS-T's implementation uses [Intel® Neural Compressor](https://github.com/intel/neural-compressor) as an underlying backend for quantizing models. This search approach is specific to the CPU, and so `--device=cpu` has to be used.

Usage example of Quantization-aware search on OFA ResNet50 super-network.

```bash
dynast \
        --results_path dynast_ofaresnet50_quant.csv \
        --dataset_path /ML_datasets/imagenet/ilsvrc12_raw \
        --supernet inc_quantization_ofa_resnet50 \
        --device cpu \
        --batch_size 128 \
        --search_tactic linas \
        --measurements latency accuracy_top1 \
        --optimization_metrics latency accuracy_top1 \
        --seed 42
```

### CODE REVIEW

File: dynast/supernetwork/image_classification/ofa_quantization/depth_parser.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        regex_module_names = []
+        for (name, module) in subnet_model.named_modules():
+            if type(module) in (nn.modules.conv.Conv2d, ) and name.endswith('.conv'):
+                regex_module_names.append(f'^({name})$')
+        return regex_module_names
+
+
+class DepthParser:
+    def __new__(self, supernet, **kwargs):
+
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/depth_parser.py:113]:
```python
if 'inc_quantization_ofa_resnet50 ` in supernet:
    ...
else:
    raise NotImplementedError('Unsupported supernet/quantization setting..')
```

File: dynast/supernetwork/image_classification/ofa_quantization/inc_quantization.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+import yaml
+from neural_compressor.experimental import Quantization, common
+# from dynast.utils.nn import validate_classification
+
+
+def default_policy():
+    policy = {
+        'model': {
+            'name': 'dynast_quantized_subnet',
+            'framework': 'pytorch_fx'
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/inc_quantization.py:15]: Maybe default should be Neural Compressor (in general, and also given the file name)?

File: dynast/supernetwork/image_classification/ofa_quantization/inc_quantization.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        'asymmetric': 'asym'
+    }
+    qgranularity = {
+        'pertensor': 'per_tensor',
+        'perchannel': 'per_channel'
+    }
+
+    assert observer_type in observer.keys()
+    assert bit in dtype.keys()
+    assert mode in ('symmetric', 'asymmetric')
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/inc_quantization.py:67]:
```python
assert mode in qscheme.keys()
```

File: dynast/supernetwork/image_classification/ofa_quantization/inc_quantization.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    assert granularity in ('perchannel', 'pertensor')
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/inc_quantization.py:68]:
```python
assert granularity in qgranularity.keys()
```

File: dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+        self.device = device
+        self.test_fraction = test_fraction
+        self.dataset_path = dataset_path
+        self.dataloader_workers = dataloader_workers
+        self.verbose = verbose
+        ImagenetDataProvider.DEFAULT_PATH = dataset_path
+        if supernet=='inc_quantization_ofa_resnet50':
+            self.ofa_network = ofa_model_zoo.ofa_net('ofa_resnet50', pretrained=True)
+        else:
+            self.ofa_network = ofa_model_zoo.ofa_net(supernet, pretrained=True)
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py:109]:
```python
ofa_supernet = supernet.replace('inc_quantization_', '')
self.ofa_network = ofa_model_zoo.ofa_net(ofa_supernet, pretrained=True)
```
If `ofa_supernet` does not exist the 2nd line will `raise ValueError("Not supported: %s" % net_id)`, so it should be safe.

File: dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+
+        self.run_config = ImagenetRunConfig(
+            test_batch_size=batch_size,
+            n_worker=dataloader_workers,
+        )
+        self.depth_parser = DepthParser(supernet='ofa_resnet50', supernet_depth=[2]*5, base_blocks=[2,2,4,2])
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py:115]: Reuse `ofa_supernet` and do sth smart about the `supernet_depth` and `base_blocks` or surround it with `if ofa_supernet == 'ofa_resnet50'`

File: dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+
+        self._init_data()
+
+        self.quantizer = Quantization(calibration_dataloader = self.calibration_dataloader,
+                             num_samples = 100
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py:120]: num_samples should be configurable. Add it to runner's init?

File: dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+                )
+            
+            if 'params' in self.measurements:
+                individual_results['params'] = self.evaluator.validate_params(
+                    subnet_sample
+                )
+            if 'latency' in self.measurements:
+                individual_results['latency'], _ = self.evaluator.measure_latency(
+                    subnet_sample
+                )  # TODO(change batch size!!)
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py:345]: todo needs to be addressed.

File: dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+        """Measure Torch model's FLOPs/MACs as per FVCore calculation
+        Args:
+            subnet_cfg: sub-network Torch model
+            device: Target device. If not provided will use self.device.
+        Returns:
+            `params`
+        """
+        time_stamp = str(datetime.now())
+
+        temp_name = f'temp_{time_stamp}'
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py:227]: This potentially can (albeit probability is low, as far as I remember now() returns time in nanosecond-level precision) when ran in distributed mode. Maybe random string instead?

```python
import uuid
temp_name = f'temp_{str(uuid.uuid4())[:8]}'

File: dynast/supernetwork/supernetwork_registry.py [hunk 10]
```diff
@@ -18,6 +18,16 @@
     EvaluationInterfaceOFAMobileNetV3,
     EvaluationInterfaceOFAResNet50,
 )
+from dynast.supernetwork.image_classification.ofa.ofa_encoding import OFAMobileNetV3Encoding, OFAResNet50Encoding
```
Comment (macsz) [inline dynast/supernetwork/supernetwork_registry.py:21]: Duplicate of line 21

File: dynast/supernetwork/supernetwork_registry.py [hunk 11]
```diff
@@ -18,6 +18,16 @@
...
+from dynast.supernetwork.image_classification.ofa.ofa_interface import (
+    EvaluationInterfaceOFAMobileNetV3,
+    EvaluationInterfaceOFAResNet50,
+)
```
Comment (macsz) [inline dynast/supernetwork/supernetwork_registry.py:25]: Ditto 17

**Review by macsz — COMMENTED** [review]
Additional comments:
- [ ] Add `neural-compressor` to requirements.txt. We should not pin version, though, as Neural Compressor installs DyNAS-T, and it could lead to some weird stuff if we did.

File: dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+        )
+        run_manager.reset_running_statistics(net=subnet)
+
+        subnet_qt = self.quantize_and_calibrate(subnet, subnet_cfg)
+
+        # Test sampled subnet
+        self.run_config.data_provider.assign_active_img_size(subnet_cfg['r'][0])
+
+        loss, top1, top5 = validate_classification(
+            model=subnet,
```
Comment (macsz) [inline dynast/supernetwork/image_classification/ofa_quantization/quantization_interface.py:201]: !!

File: dynast/supernetwork/image_classification/ofa_quantization/inc_quantization.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+    time_stamp = time.time()
+    temp_yaml_name = f'temp_{time_stamp}.yaml'
+    with open(temp_yaml_name, 'w') as f:
+        yaml.dump(qconfig_dict, f)
+    try:
+        quantizer = Quantization(temp_yaml_name)
+        quantizer.model = copy.deepcopy(model_fp)
+        quantizer.calib_dataloader = data_loader
+        # quantizer.eval_func = lambda model: validate_classification(model, data_loader=data_loader, test_size=mp_calibration_samples//data_loader.batch_size)[1]
+        model_qt = quantizer.fit()
```
Comment (sharathns93) [inline dynast/supernetwork/image_classification/ofa_quantization/inc_quantization.py:247]: remove commented code

**Comment by macsz** [comment]
Approved by Sharath and myself.

**Pull request merged**